src/wifi/model/periodic-dmg-wifi-scheduler.cc

```diff
@@ -0,0 +1,539 @@
+      m_remainingDtiTime = 0;
+      for (const auto & slot: m_availableSlots)
+        {
+          m_remainingDtiTime += (slot.second - slot.first + 1);
+        }
+    }
+}
+
+void
+PeriodicDmgWifiScheduler::AdjustExistingAllocations (AllocationFieldListI iter, uint32_t duration, bool isToAdd)
+{
+  NS_LOG_FUNCTION (this << duration << isToAdd);
+
+  // This method is called upon a DelTsRequest or after the cleanup of 
+  // non-pseudostatic allocations.
+  // In this version of the periodic scheduler, existing allocations are not shifted 
+  // to fill the created gaps but only the vector listing the available slots is updated.
+  // For this reason, the current input parameters are useless.
+
+  auto addtsListCopy = m_addtsAllocationList;
+
+  // sort the copy to simplify the process of going through the allocation list
+  sort (addtsListCopy.begin (),
+        addtsListCopy.end (),
+        [](const AllocationField& lhs, const AllocationField& rhs){
+      return lhs.GetAllocationStart () < rhs.GetAllocationStart ();
+    });
+
+  uint32_t startAlloc, endAlloc;
+
+  // clear m_availableSlots and refill it based on the updated m_addtsAllocationList
+  m_availableSlots.clear ();
+  m_availableSlots.push_back (std::make_pair (0, m_dtiDuration.GetMicroSeconds ()));
+
+  for (const auto & allocation: addtsListCopy)
+    {
+      startAlloc = allocation.GetAllocationStart ();
+      endAlloc = startAlloc + allocation.GetAllocationBlockDuration () + m_guardTime;
+      if (allocation.GetNumberOfBlocks () > 1)
+        {
+          // if the number of blocks allocated is > 1, the allocation is periodic and must be dealt with accordingly
+          uint8_t blocks = allocation.GetNumberOfBlocks ();
+          while (blocks > 0)
+            {
+              UpdateAvailableSlots (startAlloc, endAlloc);
+              startAlloc += allocation.GetAllocationBlockPeriod ();
+              endAlloc += allocation.GetAllocationBlockPeriod ();
+              // AllocationBlockPeriod represents the time between the start of two consecutive time blocks belonging to the same allocation
+              blocks -= 1;
+            }
+        }
+      else
+        {
+          UpdateAvailableSlots (startAlloc, endAlloc);
+        }
+    }
+}
+
+uint32_t
+PeriodicDmgWifiScheduler::GetAllocationDuration (uint32_t minAllocation, uint32_t maxAllocation)
+{
+  NS_LOG_FUNCTION (this << minAllocation << maxAllocation);
+  return maxAllocation;
+}
+
+StatusCode
+PeriodicDmgWifiScheduler::AddNewAllocation (uint8_t sourceAid, const DmgTspecElement &dmgTspec, const DmgAllocationInfo &info)
+{
+  NS_LOG_FUNCTION (this);
+
+  StatusCode status;
+  uint32_t allocDuration;
+
+  if (m_availableSlots.begin () == m_availableSlots.end ())
+    {
+      NS_LOG_DEBUG ("There are no free available slots in the DTI.");
+      status.SetFailure ();
+      return status;
+    }
+
+  if (info.GetAllocationFormat () == ISOCHRONOUS)
```

Comment:
I don't get what you mean, anyway! How could this be improved by scoped enum?

Reply:
https://stackoverflow.com/questions/18335861/why-is-enum-class-preferred-over-plain-enum

You would use `AllocationFormat::ISOCHRONOUS` instead of `ISOCHRONOUS`. It's safer to use and avoids some common bugs/programming mistakes, especially for projects this large